### PR TITLE
[CatalogPromotions][API] Reapply promotions after action removal

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         | master 
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Listener/CatalogPromotionConfigurationRemovedListener.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        $this->catalogPromotionRepository = $catalogPromotionRepository;
+    }
+
+    public function __invoke(CatalogPromotionConfigurationRemoved $event): void
+    {
+        $this->catalogPromotionClearer->clear();
+
+        foreach ($this->catalogPromotionRepository->findAll() as $catalogPromotion) {
+            $this->catalogPromotionProcessor->process($catalogPromotion);
+        }
```
Comment (arti0090) [inline src/Sylius/Bundle/CoreBundle/Listener/CatalogPromotionConfigurationRemovedListener.php:45]: As this is the same as in other Listener maybe I will make CatalogPromotionReapplicator class so code won't be duplicated?

File: src/Sylius/Bundle/CoreBundle/Processor/CatalogPromotionReprocessor.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Processor;
+
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+
+class CatalogPromotionReprocessor implements CatalogPromotionReprocessorInterface
```
Comment (arti0090) [inline src/Sylius/Bundle/CoreBundle/Processor/CatalogPromotionReprocessor.php:18]: This name does not resonate with me has anyone got better naming for it (also namespace is it ok?)

Comment (Arminek) [reply src/Sylius/Bundle/CoreBundle/Processor/CatalogPromotionReprocessor.php:18]: How about `CatalogPromotionProcessor`?

Comment (arti0090) [reply src/Sylius/Bundle/CoreBundle/Processor/CatalogPromotionReprocessor.php:18]: We have `CatalogPromotionProcessor` in this namespace - my class is just an extension for it cause it first clears promotions and then process them all again

File: src/Sylius/Bundle/CoreBundle/EventListener/CatalogPromotionEventListener.php [hunk 3]
```diff
@@ -49,4 +50,13 @@ public function postUpdate(LifecycleEventArgs $args): void
             $this->eventBus->dispatch(new CatalogPromotionUpdated($entity->getCatalogPromotion()->getCode()));
         }
     }
+
+    public function postRemove(LifecycleEventArgs $args): void
+    {
+        $entity = $args->getObject();
+
+        if ($entity instanceof CatalogPromotionActionInterface) {
+            $this->eventBus->dispatch(new CatalogPromotionConfigurationRemoved());
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/EventListener/CatalogPromotionEventListener.php:59]: Why didn't you use an existing event `CatalogPromotionUpdated`? Removing configuration is also an update of catalog promotion, and what is more, It looks like the behaviour in listeners is the same 🤔

File: src/Sylius/Bundle/ApiBundle/EventSubscriber/CatalogPromotionEventSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $entity = $event->getControllerResult();
+        $method = $event->getRequest()->getMethod();
+
+        if ($method === Request::METHOD_PUT || $method === Request::METHOD_PATCH) {
+            if ($entity instanceof CatalogPromotionInterface) {
+                $this->eventBus->dispatch(new CatalogPromotionUpdated($entity->getCode()));
+            }
+
+            if ($entity instanceof CatalogPromotionActionInterface) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/EventSubscriber/CatalogPromotionEventSubscriber.php:52]: Is this if needed? I guess that after updating a catalog promotion action on catalog promotion endpoint, we need to subscribe only on changes on `CatalogPromotionInterface`

File: src/Sylius/Bundle/CoreBundle/Processor/CatalogPromotionReprocessorInterface.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Processor;
+
+interface CatalogPromotionReprocessorInterface
+{
+    public function reprocess(): void;
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Processor/CatalogPromotionReprocessorInterface.php:18]:
```suggestion
    public function process(): void;
```
?

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, @arti0090! :1st_place_medal:

**Pull request merged**